R/future_promise.R

```diff
@@ -281,7 +300,194 @@ future_promise_queue <- local({
+#'   # We should NOT expect to see `print_every()` statements
+#'   #   while `future::future()` is blocking the main R session
+#'   print_every <- function(i = 0, max = 5 / delay, delay = 0.25) {
+#'     if (i > max) return()
+#'     cat_ex("print_every(): ", i, "/", max)
+#'     # Print again, later
+#'     later::later(function() { print_every(i + 1, max = max, delay = delay) }, delay = delay)
+#'   }
+#'
+#'   # Act as if the main R session is free for the next `timeout` seconds
+#'   run_ex_now <- function() {
+#'     while (!later::loop_empty()) {
+#'       later::run_now()
+#'       Sys.sleep(0.01)
+#'     }
+#'     message("done with run_ex_now()")
+#'   }
+#'
+#'   # Consistent printing method for the example below
+#'   print_ex <- function(x) {
+#'     promise_all(.list = x) %...>%
+#'       {
+#'         cat("\n")
+#'         print(.);
+#'         cat("\nlog:\n", paste0(readLines(log), collapse = "\n"), "\n", sep = "")
+#'       }
+#'   }
+#'
+#'
+#'   ## Example ##
+#'
+#'   # Execute 10 `future_promise()` calls.
+#'   # In this example, we expect
+#'   # * all _creating_ statements should appear immediately
+#'   # * the first two `future_promise()` calls should start immediately
+#'   # * two `future_promise()` calls should execute simultaneously throughout the process
+#'   # * the log should have `print_every()` statements interleaved with `future_promise()` statements
+#'   {
+#'     log <- tempfile(); start <- Sys.time()
+#'     print_every()
+#'     lapply(1:10, function(i) {
+#'       cat_ex("future_promise(): ", i, " creating")
+#'       future_promise({
+#'         cat_ex("future_promise(): ", i, " starting")
+#'         Sys.sleep(1)
+#'         cat_ex("future_promise(): ", i, " finished")
+#'         i
+#'       })
+#'     }) %>%
+#'       print_ex()
+#'     cat_ex("done with lapply()")
+#'
+#'     run_ex_now() # ~5s; max(5s for `future_promise()`, 5s for `print_every()`)
+#'   }
+#'
+#'   ## Unexpected behavior - Blocking the main R session ##
+#'
+#'   # Execute 10 `future::future()` calls.
+#'   # In this example, we expect
+#'   # * the two _creating_ statements should appear immediately.
+#'   #     The remaining _creating_ statements will appear after prior work has completed
+#'   # * the first two `future::future()` calls should start immediately
+#'   # * two `future::future()` calls should execute simultaneously throughout the process
+#'   # * the log will NOT have `print_every()` statements mixed with `future::future()` statements.
+#'   #     This means the main R session is blocked by `future` waiting for a worker to become free
+#'   {
+#'     log <- tempfile(); start <- Sys.time()
+#'     print_every()
+#'     lapply(1:10, function(i) {
+#'       cat_ex("future::future(): ", i, " creating")
+#'       future::future({
+#'         cat_ex("future::future(): ", i, " starting")
+#'         Sys.sleep(1)
+#'         cat_ex("future::future(): ", i, " finished")
+#'         i
+#'       })
```

Comment:
Let's put this example before the `future_promise()` example to demo the problem right away